### ISSUE TITLE
Flathubbot before merging a build is complete

### ISSUE THREAD

**Issue by Pobega** [flathub-infra/flatpak-external-data-checker#167]
https://github.com/flathub/com.fightcade.Fightcade/pull/38

It seems that after a build fails the trigger for Flathubbot merging is broken. In my case a build fails because I upstreamed a patch, when I pushed the removal of that patch Flathubbot correctly triggered the next test build but automerged my PR before the build was complete.

I'm unsure if this is intended behavior, if it is feel free to close this.

If this is the wrong repository for Flathubbot issues please point me in the right direction, I couldn't find a repository for Flathubbot specifically.

Here's a screenshot to show the context in case it becomes unclear after the build passes.

![image](https://user-images.githubusercontent.com/474525/115238521-c783e280-a0eb-11eb-9138-47aa8fde619e.png)

**Comment by gasinvein** [comment]
This surely isn't intended behavior, we explicitly do a check if build was successful before merging. The check is based on GitHub CI status of PR, so this looks like an unfortunate coincidence - probably f-e-d-c found the PR before CI status was updated to "running" after successful sources download phase.

**Comment by gasinvein** [comment]
Here is the log line:
```
+ 2021-04-19 12:43:58,760    INFO src.main: PR passed CI and is mergeable, merging https://github.com/flathub/com.fightcade.Fightcade/pull/38
```
It comes from here:
https://github.com/flathub/flatpak-external-data-checker/blob/c5c89339d808c68b4d47088c3998ce1d386b0111/src/main.py#L224-L227

So, to my understanding, the only way for this to happen could've been that the PR CI status was marked as successful at the moment f-e-d-c was checking it.

This is unlikely but serious issue, we should think about preventing it. @barthalion, can you please educate me when and how PR CI status is updated during flatpak build?

**Comment by Pobega** [comment]
Due to the timing I'm curious if it's related to the `download-sources` CI passing (since the CI seems to be split into both a `download-sources` and `builds/x86_64`)

**Comment by gasinvein** [comment]
Yes, I believe it is. Looks like there is some time frame between finishing sources download and starting builds, during which CI status is already set to "success" but not yet changed back to "running".

**Comment by wjt** [comment]
Could the bot explicitly check that there is at least one `builds/*` jobs and all `builds/*` jobs have succeeded?

**Comment by gasinvein** [comment]
Seems a bit hackish but doable. Yet I'd like to avoid flathub-specific code, if possible.

**Comment by barthalion** [comment]
I guess checking if any job has actually run would be enough here, no need to make it specific to Flathub.

**Comment by gasinvein** [comment]
Such check will behave same as the current one: sources download job already completed, build jobs not started yet, so there are only completed jobs and the check will pass.

**Comment by barthalion** [comment]
Ultimately, I'd say that there's nothing wrong about Flathub-specific code given Flathub is primary "customer" of the tool. An alternative would be to require status checks to pass across all repos:

![image](https://user-images.githubusercontent.com/380403/117449886-c999ce00-af40-11eb-9d23-e2b8691d5b50.png)

…which I believe would ensure the PR is only mergeable when chosen checks passed.

**Comment by bbhtt** [comment]
This should not be an issue anymore.
